File: CSharp/Urho3D/CSharp/Math/Rect.cs
using System;
using System.Runtime.InteropServices;

namespace Urho3D
{
	[StructLayout(LayoutKind.Sequential)]
	public struct Rect
	{
        /// Minimum vector.
        public Vector2 Min;
        /// Maximum vector.
        public Vector2 Max;

		public Rect(float minX, float minY, float maxX, float maxY)
		{
			Min = new Vector2(minX, minY);
			Max = new Vector2(maxX, maxY);
		}

	    /// Merge a point.
	    public void Merge(Vector2 point)
	    {
	        if (point.X < Min.X)
	            Min.X = point.X;
	        if (point.X > Max.X)
	            Max.X = point.X;
	        if (point.Y < Min.Y)
	            Min.Y = point.Y;
	        if (point.Y > Max.Y)
	            Max.Y = point.Y;
	    }

	    /// Merge a rect.
	    public void Merge(Rect rect)
	    {
	        if (rect.Min.X < Min.X)
	            Min.X = rect.Min.X;
	        if (rect.Min.Y < Min.Y)
	            Min.Y = rect.Min.Y;
	        if (rect.Max.X > Max.X)
	            Max.X = rect.Max.X;
	        if (rect.Max.Y > Max.Y)
	            Max.Y = rect.Max.Y;
	    }

		public static readonly Rect Full = new Rect(-1.0f, -1.0f, 1.0f, 1.0f);
		public static readonly Rect Positive = new Rect(0.0f, 0.0f, 1.0f, 1.0f);
		public static readonly Rect Zero = new Rect(0.0f, 0.0f, 0.0f, 0.0f);
	}

	[StructLayout(LayoutKind.Sequential)]
	public struct IntRect
	{
        /// Left coordinate.
        public int Left;
        /// Top coordinate.
        public int Top;
        /// Right coordinate.
        public int Right;
        /// Bottom coordinate.
        public int Bottom;

		public IntRect(int left, int top, int right, int bottom)
		{
			Left = left;
			Top = top;
			Right = right;
			Bottom = bottom;
		}

		public static readonly IntRect Zero = new IntRect(0, 0, 0, 0);
	}
}

File: Source/Urho3D/Math/Rect.cpp
// Copyright (c) 2008-2022 the Urho3D project.
// Copyright (c) 2022-2026 the rbfx project.
// This work is licensed under the terms of the MIT license.
// For a copy, see <https://opensource.org/licenses/MIT> or the accompanying LICENSE file.

#include "Urho3D/Precompiled.h"

#include "Urho3D/Math/Rect.h"

#include "Urho3D/Container/Str.h"

#include <cstdio>

#include "Urho3D/DebugNew.h"

namespace Urho3D
{

const Rect Rect::FULL(-1.0f, -1.0f, 1.0f, 1.0f);
const Rect Rect::POSITIVE(0.0f, 0.0f, 1.0f, 1.0f);
const Rect Rect::ZERO(0.0f, 0.0f, 0.0f, 0.0f);

const IntRect IntRect::ZERO(0, 0, 0, 0);

void IntRect::Clip(const IntRect& rect)
{
    if (rect.left_ > left_)
        left_ = rect.left_;
    if (rect.right_ < right_)
        right_ = rect.right_;
    if (rect.top_ > top_)
        top_ = rect.top_;
    if (rect.bottom_ < bottom_)
        bottom_ = rect.bottom_;

    if (left_ >= right_ || top_ >= bottom_)
        *this = IntRect();
}

void IntRect::Merge(const IntRect& rect)
{
    if (Width() <= 0 || Height() <= 0)
    {
        *this = rect;
    }
    else if (rect.Width() > 0 && rect.Height() > 0)
    {
        if (rect.left_ < left_)
            left_ = rect.left_;
        if (rect.top_ < top_)
            top_ = rect.top_;
        if (rect.right_ > right_)
            right_ = rect.right_;
        if (rect.bottom_ > bottom_)
            bottom_ = rect.bottom_;
    }
}

ea::string Rect::ToString() const
{
    char tempBuffer[CONVERSION_BUFFER_LENGTH];
    snprintf(tempBuffer, CONVERSION_BUFFER_LENGTH, "%g %g %g %g", min_.x_, min_.y_, max_.x_, max_.y_);
    return ea::string(tempBuffer);
}

ea::string IntRect::ToString() const
{
    char tempBuffer[CONVERSION_BUFFER_LENGTH];
    snprintf(tempBuffer, CONVERSION_BUFFER_LENGTH, "%d %d %d %d", left_, top_, right_, bottom_);
    return ea::string(tempBuffer);
}

void Rect::Clip(const Rect& rect)
{
    if (rect.min_.x_ > min_.x_)
        min_.x_ = rect.min_.x_;
    if (rect.max_.x_ < max_.x_)
        max_.x_ = rect.max_.x_;
    if (rect.min_.y_ > min_.y_)
        min_.y_ = rect.min_.y_;
    if (rect.max_.y_ < max_.y_)
        max_.y_ = rect.max_.y_;

    if (min_.x_ > max_.x_ || min_.y_ > max_.y_)
    {
        min_ = Vector2(M_INFINITY, M_INFINITY);
        max_ = Vector2(-M_INFINITY, -M_INFINITY);
    }
}

}
